internal/cloud/realm/client.go

```diff
@@ -91,5 +91,56 @@ func (c *client) do(method, path string, options api.RequestOptions) (*http.Resp
 		req.Header.Set(api.HeaderAuthorization, "Bearer "+auth)
 	}
 	client := &http.Client{}
-	return client.Do(req)
+	res, err := client.Do(req)
+	if err != nil {
+		return res, err
+	}
+	if res.StatusCode != http.StatusUnauthorized || !options.UseAuth || options.PreventRefresh {
+		return res, err
+	}
+
+	serverError := unmarshalServerError(res).(ServerError)
+	if serverError.Code != InvalidSessionCode {
+		return res, err
+	}
+
+	authToken, refreshErr := c.refreshAuth(client)
+	if refreshErr != nil {
+		return nil, refreshErr
+	}
+	// TODO: REALMC-7719 save the new access token to prevent unnecessary retries
+	c.session.AccessToken = authToken
+	options.PreventRefresh = true
+
+	return c.do(method, path, options)
+}
+
+func (c *client) refreshAuth(httpClient *http.Client) (string, error) {
```

Comment:
open to changing the signature or moving this function into the auth package. I felt this function was for internal client behavior so I left it here and re-used the httpClient

Reply:
could this refreshAuth/refreshSession method implementation leverage `func (c *client) do` and take care of a lot of this code by just specifying the proper `RequestOptions`?